word_segmentation.py

```diff
@@ -286,35 +293,287 @@ def preprocess():
+            new_line += "\n"
+            wfile.write(new_line)
+
+
+def compute_ICU_accuracy(filename):
+    """
+    This function uses a dataset to compute the accuracy of icu word breakIterator
+    Args:
+        filename: The path of the file
+    """
+    chars_break_iterator = BreakIterator.createCharacterInstance(Locale.getUS())
+    line_counter = 0
+    icu_mismatch = 0
+    icu_total_bies_lengths = 0
+    with open(filename) as f:
+        for line in f:
+            line = clean_line(line)
+            if line == -1:
+                continue
+
+            # Finding word breakpoints using the segmented data
+            word_brkpoints = []
+            found_bars = 0
+            for i in range(len(line)):
+                if line[i] == '|':
+                    word_brkpoints.append(i - found_bars)
+                    found_bars += 1
+
+            # Creating the unsegmented line
+            unsegmented_line = line.replace("|", "")
+
+            # Making the grapheme clusters brkpoints
+            chars_break_iterator.setText(unsegmented_line)
+            char_brkpoints = [0]
+            for brkpoint in chars_break_iterator:
+                char_brkpoints.append(brkpoint)
+            true_bies = get_bies(char_brkpoints, word_brkpoints)
+            true_bies_str = get_bies_string_from_softmax(np.transpose(true_bies))
+
+            # Compute segmentations of icu and BIES associated with it
+            words_break_iterator = BreakIterator.createWordInstance(Locale.getUS())
+            words_break_iterator.setText(unsegmented_line)
+            icu_word_brkpoints = [0]
+            for brkpoint in words_break_iterator:
+                icu_word_brkpoints.append(brkpoint)
+            icu_bies = get_bies(char_brkpoints, icu_word_brkpoints)
+            icu_bies_str = get_bies_string_from_softmax(np.transpose(icu_bies))
+
+            # Counting the number of mismatches between icu_bies and true_bies
+            icu_total_bies_lengths += len(icu_bies_str)
+            icu_mismatch += diff_strings(true_bies_str, icu_bies_str)
+
+            line_counter += 1
+        icu_accuracy = 1 - icu_mismatch / icu_total_bies_lengths
+        return icu_accuracy
+
+
+def divide_train_test_data(input_text, train_text, valid_text, test_text):
+    """
+    This function divides a file into three new files, that contain train data, validation data, and testing data
+    Args:
+        input_text: address of the original file
+        train_text: address to store the train data in it
+        valid_text: address to store the validation data in it
+        test_text: address to store the test file in it
+    """
+    train_ratio = 0.4
+    valid_ratio = 0.4
+    train_file = open(train_text, 'w')
+    valid_file = open(valid_text, 'w')
+    test_file = open(test_text, 'w')
+    num_lines = sum(1 for _line in open(input_text))
+    line_counter = 0
+    with open(input_text) as f:
+        for line in f:
+            line_counter += 1
```

Comment:
Hmm.  Some of the data files have high locality, in the sense that there might be more news articles toward the top and religious texts toward the bottom.  I would be more comfortable if we could randomize this across the whole file, or shuffle the lines before bucketing them.

Reply:
Good point! However, I think by randomizing all lines we may lose context that the LSTM model may be able to learn. What I do is that I put every `bucket_size = 20` lines into one group, and then write the first 8 lines in `train_file`, the next 8 lines in `valid_file`, and the remaining in the `test_file`.